In [0]:
# imports
from pyspark.sql import SparkSession
import pyspark.sql.functions as f
import pyspark.sql.types as t

In [0]:
spark = SparkSession.builder.master("yarn").appName("MovieAnalysis").getOrCreate()

In [0]:
# schemes to be defined
from pyspark.sql.types import *
from pyspark.sql.functions import *

lang_schema = StructType([StructField('name', StringType(), True),
                          StructField('id', IntegerType(), True)])

cast_schema = StructType([StructField('name', StringType(), True),
                          StructField('character', StringType(), True),
                          StructField('gender', IntegerType(), True),
                          StructField('id', IntegerType(), True)])

genre_schema = StructType([StructField('name', StringType(), True),
                          StructField('id', IntegerType(), True)])

movie_schema = StructType([StructField('index', IntegerType(), True),
                          StructField('title', IntegerType(), True),
                          StructField('release_date', DateType(), True),
                          StructField('runtime', FloatType(), True),
                          StructField('revenue', IntegerType(), True),
                          StructField('budget', IntegerType(), True),
                          StructField('popularity', FloatType(), True),
                          StructField('id', IntegerType(), True)])

crew_schema = StructType([StructField('name', StringType(), True),
                          StructField('job', StringType(), True),
                          StructField('gender', IntegerType(), True),
                          StructField('id', IntegerType(), True)])

recom_schema = StructType([StructField('index', IntegerType(), True),
                          StructField('movie_id', IntegerType(), True),
                          StructField('user_id', IntegerType(), True),
                          StructField('vote', IntegerType(), True)])

In [0]:
# maybe add schemes when reading the csv files
df_lang = spark.read.format("csv").option("header", "true").load("dbfs:/FileStore/shared_uploads/ibele@stud.dhbw-ravensburg.de/spoken_languages-1.csv")
df_cast = spark.read.format("csv").option("header", "true").load("dbfs:/FileStore/shared_uploads/ibele@stud.dhbw-ravensburg.de/cast-3.csv")
df_genre = spark.read.format("csv").option("header", "true").load("dbfs:/FileStore/shared_uploads/ibele@stud.dhbw-ravensburg.de/genres-2.csv")
df_movie = spark.read.format("csv").option("header", "true").load("dbfs:/FileStore/shared_uploads/ibele@stud.dhbw-ravensburg.de/movies-3.csv")
df_crew = spark.read.format("csv").option("header", "true").load("dbfs:/FileStore/shared_uploads/ibele@stud.dhbw-ravensburg.de/crew-2.csv")
df_recom = spark.read.format("csv").option("header", "true").load("dbfs:/FileStore/shared_uploads/ibele@stud.dhbw-ravensburg.de/recom.csv")

In [0]:
# dropping first column of movie since it's only the index and printing the schema
df_movie = df_movie.drop("_c0")
df_movie.printSchema()

root
 |-- title: string (nullable = true)
 |-- release_date: string (nullable = true)
 |-- runtime: string (nullable = true)
 |-- revenue: string (nullable = true)
 |-- budget: string (nullable = true)
 |-- popularity: string (nullable = true)
 |-- id: string (nullable = true)



In [0]:
df_movie.display()

title release_date runtime revenue budget popularity id Avatar 2009-12-10 162.0 2787966824 236972821 150.437577 19995 Pirates of the Caribbean: At World's End 2007-05-19 169.0 961022070 299983825 139.082615 285 Spectre 2015-10-26 148.0 880719915 244983844 107.376788 206647 The Dark Knight Rises 2012-07-16 165.0 1084987218 249951643 112.31295 49026 John Carter 2012-03-07 132.0 284133596 260013038 43.926995 49529 Spider-Man 3 2007-05-01 139.0 890863843 258008881 115.699814 559 Tangled 2010-11-24 100.0 591814703 260023415 48.681969 38757 Avengers: Age of Ultron 2015-04-22 141.0 1405364340 279979533 134.279229 99861 Harry Potter and the Half-Blood Prince 2009-07-07 153.0 933974783 250021729 98.885637 767 Batman v Superman: Dawn of Justice 2016-03-23 151.0 873305820 249958606 155.790452 209112 Superman Returns 2006-06-28 154.0 391115123 270030875 57.925623 1452 Quantum of Solace 2008-10-30 106.0 586093018 200046124 107.928811 10764 Pirates of the Caribbean: Dead Man's Chest 2006-06-20 151.0 1065634246 200043847 145.847379 58 The Lone Ranger 2013-07-03 149.0 89259867 254963414 49.046956 57201 Man of Steel 2013-06-12 143.0 662802352 224971303 99.398009 49521 The Chronicles of Narnia: Prince Caspian 2008-05-15 150.0 419645568 224990122 53.978602 2454 The Avengers 2012-04-25 143.0 1519574823 219978563 144.448633 24428 Pirates of the Caribbean: On Stranger Tides 2011-05-14 136.0 1045723650 380000081 135.413856 1865 Men in Black 3 2012-05-23 106.0 623999006 224983577 52.035179 41154 The Hobbit: The Battle of the Five Armies 2014-12-10 144.0 955973335 249996564 120.965743 122917 The Amazing Spider-Man 2012-06-27 136.0 752171797 215011329 89.866276 1930 Robin Hood 2010-05-12 140.0 310709540 200021407 37.668301 20662 The Hobbit: The Desolation of Smaug 2013-12-11 161.0 958376165 249956828 94.370564 57158 The Golden Compass 2007-12-04 113.0 372216307 180022161 42.990906 2268 King Kong 2005-12-14 187.0 549995282 206955001 61.22601 254 Titanic 1997-11-18 194.0 1845063883 199965341 100.025899 597 Captain America: Civil War 2016-04-27 147.0 1153255837 250005667 198.372395 271110 Battleship 2012-04-11 131.0 302983618 208974208 64.928382 44833 Jurassic World 2015-06-09 124.0 1513518045 150023243 418.708552 135397 Skyfall 2012-10-25 143.0 1108547997 199997023 93.004993 37724 Spider-Man 2 2004-06-25 127.0 783788409 199959277 35.149586 558 Iron Man 3 2013-04-18 130.0 1215448495 200030236 77.68208 68721 Alice in Wonderland 2010-03-03 108.0 1025503271 199981857 78.530105 12155 X-Men: The Last Stand 2006-05-24 104.0 459400259 210047540 3.857526 36668 Monsters University 2013-06-20 104.0 743589696 200004961 89.186492 62211 Transformers: Revenge of the Fallen 2009-06-19 150.0 836282582 149957721 21.939663 8373 Transformers: Age of Extinction 2014-06-25 165.0 1091412035 209953426 116.840296 91314 Oz: The Great and Powerful 2013-03-07 130.0 491880227 199991047 46.985445 68728 The Amazing Spider-Man 2 2014-04-16 142.0 705762012 199966897 89.270217 102382 TRON: Legacy 2010-12-10 125.0 400107062 170000549 73.79505 20526 Cars 2 2011-06-11 106.0 559873703 200005403 49.98659 49013 Green Lantern 2011-06-16 114.0 219897841 200014075 51.872839 44912 Toy Story 3 2010-06-16 103.0 1066982178 199951160 59.995418 10193 Terminator Salvation 2009-05-20 115.0 371311820 200011306 71.862892 534 Furious 7 2015-04-01 137.0 1506286614 189962981 102.322217 168259 World War Z 2013-06-20 116.0 531840685 199986059 81.834855 72190 X-Men: Days of Future Past 2014-05-15 131.0 747894016 250035421 118.078691 127585 Star Trek Into Darkness 2013-05-05 132.0 467349228 189955400 78.291018 54138 Jack the Giant Slayer 2013-02-27 114.0 197704496 195032996 43.349855 81005 The Great Gatsby 2013-05-10 143.0 351006583 104983057 61.196071 64682 Prince of Persia: The Sands of Time 2010-05-19 116.0 335129531 149967536 62.169881 9543 Pacific Rim 2013-07-11 131.0 407645331 179970045 56.523205 68726 Transformers: Dark of the Moon 2011-06-28 154.0 1123772152 195039732 28.529607 38356 Indiana Jones and the

In [0]:
# dropping first column of recom since it's only the index and printing the schema
df_recom = df_recom.drop("_c0")
df_recom.printSchema()

root
 |-- movie_id: string (nullable = true)
 |-- user_id: string (nullable = true)
 |-- vote: string (nullable = true)



In [0]:
# writing the dataframes to parquet
df_lang.write.format("parquet").mode("overwrite").save("output/lang.parquet")
df_cast.write.format("parquet").mode("overwrite").save("output/cast.parquet")
df_genre.write.format("parquet").mode("overwrite").save("output/genre.parquet")
df_movie.write.format("parquet").mode("overwrite").save("output/movie.parquet")
df_crew.write.format("parquet").mode("overwrite").save("output/crew.parquet")
df_recom.write.format("parquet").mode("overwrite").save("output/recom.parquet")

In [0]:
# reading the data from the parquet files in dataframes
pf_lang = spark.read.parquet("/output/lang.parquet")
pf_cast = spark.read.parquet("/output/cast.parquet")
pf_genre = spark.read.parquet("/output/genre.parquet")
pf_movie = spark.read.parquet("/output/movie.parquet")
pf_crew = spark.read.parquet("/output/crew.parquet")
pf_recom = spark.read.parquet("/output/recom.parquet")

<b> Question: </b>\
How many movies were written by a female writer?\
Explain what data storage structure you used to store the information and why. When storing the information how can you speed up the information retrieval if you know you are interested in looking at the gender of the writer? Why does it speed up the information retrieval when you store the data differently?

In [0]:
(pf_movie
     .join(pf_crew, on=pf_movie.id==pf_crew.id.astype(t.IntegerType()))
     .where(
         (pf_crew.job=="Writer") 
         & (pf_crew.gender==1))
     .orderBy("name")
     .count())

Out[36]: 151

In [0]:
(pf_movie
     .join(pf_crew, on=pf_movie.id==pf_crew.id.astype(t.IntegerType()))
     .where(
         (pf_crew.job=="Writer") 
         & (pf_crew.gender==1))
     .orderBy("name")
     .display())

title release_date runtime revenue budget popularity id name job gender id Valentine's Day 2010-02-10 125.0 216533382 52017397 25.41929 32856 Abby Kohn Writer 1 32856.0 He's Just Not That Into You 2009-02-06 129.0 177253008 39970675 26.253357 10184 Abby Kohn Writer 1 10184.0 After.Life 2010-04-09 104.0 3651621 4478061 14.570359 36419 Agnieszka WojtowiczVosloo Writer 1 36419.0 Beloved 1998-10-16 172.0 8547 48177 1.453765 39437 Akosua Busia Writer 1 39437.0 Goddess of Love 2015-08-31 96.0 -31219 -20384 1.273773 347764 Alexis Kendra Writer 1 347764.0 Saving Face 2004-09-12 91.0 -44222 -5633 2.487255 19316 Alice Wu Writer 1 19316.0 Four Rooms 1995-12-09 98.0 4253036 4047190 22.87623 5 Allison Anders Writer 1 5.0 Sugar Town 1999-09-17 92.0 -2946 -46804 0.375423 142132 Allison Anders Writer 1 142132.0 Ask Me Anything 2014-04-19 100.0 -7182 941597 8.976128 271185 Allison Burnett Writer 1 271185.0 The Betrayed 2008-09-27 98.0 6148 3539607 1.348114 20055 Amanda Gusack Writer 1 20055.0 Dawn of the Planet of the Apes 2014-06-26 130.0 710639105 170014260 243.791743 119450 Amanda Silver Writer 1 119450.0 Eye for an Eye 1996-01-12 101.0 48778 20035496 3.996392 17182 Amanda Silver Writer 1 17182.0 Step Up 3D 2010-08-04 107.0 159299849 30002212 25.060342 41233 Amy Andelson Writer 1 41233.0 Love Letters 1983-04-01 98.0 1627 585695 0.001586 75986 Amy Holden Jones Writer 1 75986.0 Trainwreck 2015-07-17 125.0 140812842 34958806 72.774941 271718 Amy Schumer Writer 1 271718.0 Fish Tank 2009-05-14 123.0 2326980 3033469 12.68672 24469 Andrea Arnold Writer 1 24469.0 In the Land of Blood and Honey 2011-12-23 127.0 264654 10015115 4.267466 79777 Angelina Jolie Writer 1 79777.0 Intolerance 1916-09-04 197.0 8441825 411937 3.232447 3059 Anita Loos Writer 1 3059.0 How She Move 2008-01-25 94.0 15961 35813 1.782666 16988 Annemarie Morais Writer 1 16988.0 The Kid 2000-07-07 104.0 69728265 65018096 18.281798 4244 Audrey Wells Writer 1 4244.0 Middle of Nowhere 2012-10-12 101.0 34596 -5734 2.836537 83588 Ava DuVernay Writer 1 83588.0 Mama 2013-01-17 100.0 146502442 14967450 47.455148 132232 Barbara Muschietti Writer 1 132232.0 The East 2013-05-31 116.0 2430785 6484224 16.574756 87499 Brit Marling Writer 1 87499.0 Another Earth 2011-07-22 92.0 1730736 94149 25.402697 55420 Brit Marling Writer 1 55420.0 Sound of My Voice 2011-01-22 85.0 -593 -3630 6.192383 86812 Brit Marling Writer 1 86812.0 Miss Congeniality 2000-12-14 111.0 212016377 44965832 32.624355 1493 Caryn Lucas Writer 1 1493.0 The Mortal Instruments: City of Bones 2013-08-21 130.0 90519297 59996494 45.514031 123553 Cassandra Clare Writer 1 123553.0 Plush 2013-09-13 98.0 12720 -9677 5.099368 215881 Catherine Hardwicke Writer 1 215881.0 Dreaming of Joseph Lees 1999-10-29 92.0 -14907 1962656 0.11615 108346 Catherine Linstrum Writer 1 108346.0 A Shine of Rainbows 2009-10-01 101.0 -20652 6006901 12.099486 37206 Catherine Spear Writer 1 37206.0 Wordplay 2006-06-26 94.0 -13141 -9061 0.906096 35199 Christine OMalley Writer 1 35199.0 Aloft 2014-02-12 112.0 45410 7983322 6.500835 251321 Claudia Llosa Writer 1 251321.0 The Wedding Date 2005-02-04 88.0 47213531 15043230 10.194097 6961 Dana Fox Writer 1 6961.0 Cadillac Records 2008-12-05 109.0 -46590 12007497 13.427427 14299 Darnell Martin Writer 1 14299.0 Can't Hardly Wait 1998-06-12 100.0 25613466 9970126 11.350382 15037 Deborah Kaplan Writer 1 15037.0 Josie and the Pussycats 2001-04-06 98.0 -17453 21997031 6.404571 19366 Deborah Kaplan Writer 1 19366.0 Leap Year 2010-01-08 100.0 29957890 18968457 21.959742 25195 Deborah Kaplan Writer 1 25195.0 Chain Letter 2010-10-01 96.0 -30016 -17628 2.504156 38033 Diana Erwin Writer 1 38033.0 The Looking Glass 2015-10-23 110.0 10388 21831 0.062979 365052 Dorothy Tristan Writer 1 365052.0 Movie 43 2013-01-01 90.0 32460369 5965847 35.350303 87818 Elizabeth Wright Shapiro Writer 1 87818.0 Step Up 3D 2010-08-04 107.0 159299849 30002212 25.060342 41233 Emily Meyer Writer 1 41233.0 Sisters in Law 2005-09-03 104.0 -48826 -2730 0.025211

<b> Answer: </b>\
Under the assumption that '1' in the column gender stands for female (2 for male and 0 for the ones, where the gender is not specified), there are 151 movies, that were written by a female writer.